File: Data Structures & Algorithms/count-connected-components/submission-2.py
class Solution:
    def countComponents(self, n: int, edges: List[List[int]]) -> int:
        graph = defaultdict(list)
        for u,v in edges:
            graph[u].append(v)
            graph[v].append(u)
        visited = set()
        
        def dfs(root):
            if root in visited:
                return
            
            visited.add(root)
            
            for nei in graph[root]:
                if nei not in visited:
                    dfs(nei)
        
        count = 0
        for i in range(n):
            if i not in visited:
                dfs(i)
                count += 1
        
        return count

File: Data Structures & Algorithms/network-delay-time/submission-2.kt
class Solution {
    fun networkDelayTime(times: Array<IntArray>, n: Int, k: Int): Int {
        val graph = mutableMapOf<Int, MutableList<Pair<Int, Int>>>()
        for ((u,v,w) in times)
            graph.getOrPut(u) {mutableListOf()}.add(Pair(v, w))

        val timeCount = IntArray(n+1){Int.MAX_VALUE}
        timeCount[0] = 0
        fun dfs(root: Int, time: Int) {
            if(timeCount[root] < time)
                return

            timeCount[root] = time
            for ((nei, t) in graph[root] ?: emptyList())
                dfs(nei, time + t)
        }
        
        dfs(k, 0)
        return if (timeCount.max() == Int.MAX_VALUE) -1 else timeCount.max()

    }
}
